src/sha256.nr

```diff
@@ -512,6 +485,173 @@ fn hash_final_block(msg_block: MSG_BLOCK, mut state: STATE) -> HASH {
+ * @notice used for complex/ recursive offloading of post-partial hashing
+ *
+ * @param N - the maximum length of the message to hash
+ * @param h - the intermediate hash state
+ * @param msg - the preimage to hash
+ * @param message_size - the actual length of the preimage to hash
+ * @return the intermediate hash state after compressing in msg to h
+ */
+pub fn partial_sha256_var_interstitial<let N: u32>(
+    mut h: [u32; 8],
+    msg: [u8; N],
+    message_size: u32,
+) -> [u32; 8] {
+    assert(message_size % BLOCK_SIZE == 0, "Message size must be a multiple of the block size");
+    if std::runtime::is_unconstrained() {
+        unsafe {
+            __sha_partial_var_interstitial(h, msg, message_size)
+        }
+    } else {
+        let (mut h, _, _) = process_full_blocks(msg, message_size, N, h);
+
+        h
+    }
+}
+
+/**
+ * Given some state of a partially computed sha256 hash and remaining preimage, complete the hash
+ * @notice used for traditional partial hashing
+ *
+ * @param N - the maximum length of the message to hash
+ * @param h - the intermediate hash state
+ * @param msg - the remaining preimage to hash
+ * @param message_size - the size of the remaining preimage
+ * @param real_message_size -= the total size of the preimage
+ * @return finalized sha256 hash
+ */
+pub fn partial_sha256_var_end<let N: u32>(
+    mut h: [u32; 8],
+    msg: [u8; N],
+    message_size: u32,
+    real_message_size: u32,
+) -> [u8; 32] {
+    assert(message_size % BLOCK_SIZE == 0, "Message size must be a multiple of the block size");
+    if std::runtime::is_unconstrained() {
+        unsafe {
+            __sha_partial_var_end(h, msg, message_size, real_message_size)
+        }
+    } else {
+        let (mut h, mut msg_block, mut msg_byte_ptr) = process_full_blocks(msg, message_size, N, h);
+        finalize_sha256_blocks(msg, real_message_size, N, h, msg_block, msg_byte_ptr)
+    }
+}
+
+pub unconstrained fn __sha_partial_var_interstitial<let N: u32>(
+    mut h: [u32; 8],
+    msg: [u8; N],
+    message_size: u32,
+) -> [u32; 8] {
+    let num_full_blocks = message_size / BLOCK_SIZE;
+    // Intermediate hash, starting with the canonical initial value
+    // Pointer into msg_block on a 64 byte scale
+    for i in 0..num_full_blocks {
+        let (msg_block, _) = build_msg_block(msg, message_size, BLOCK_SIZE * i);
+        h = sha256_compression(msg_block, h);
+    }
+    h
+}
+
+pub unconstrained fn __sha_partial_var_end<let N: u32>(
+    mut h: [u32; 8],
+    msg: [u8; N],
+    message_size: u32,
+    real_message_size: u32,
+) -> [u8; 32] {
+    h = __sha_partial_var_interstitial(h, msg, message_size);
+
+    // Handle setup of the final msg block.
+    // This case is only hit if the msg is less than the block size,
+    // or our message cannot be evenly split into blocks.
+
+    finalize_last_sha256_block(h, real_message_size, msg)
+}
```

Comment:
Does this need to be public? Note that accessing this directly avoids the safety check on `N` in `partial_sha256_var_end`